### ISSUE TITLE
contourf: ValueError: A LinearRing must have at least 3 coordinate tuples

### ISSUE THREAD

**Issue by akrherz** [SciTools/cartopy#883]
I've been migrating from basemap to cartopy and a code path that works in basemap, will now error in cartopy like so:
```
Traceback (most recent call last):
  File "current_bias.py", line 111, in <module>
    main()
  File "current_bias.py", line 108, in main
    doit(ts, model)
  File "current_bias.py", line 76, in doit
    m.contourf(lons, lats, vals, range(-10, 11, 2), units='F', cmap=cmap)
  File "/opt/miniconda2/lib/python2.7/site-packages/pyiem/plot/geoplot.py",
line 863, in contourf
    extend='both', transform=ccrs.PlateCarree())
  File "/opt/miniconda2/lib/python2.7/site-packages/cartopy/mpl/geoaxes.py",
line 1339, in contourf
    if col.get_paths()])
  File
"/opt/miniconda2/lib/python2.7/site-packages/matplotlib/collections.py",
line 213, in get_datalim
    paths = [transform.transform_path_non_affine(p) for p in paths]
  File
"/opt/miniconda2/lib/python2.7/site-packages/matplotlib/transforms.py", line
2386, in transform_path_non_affine
    return self._a.transform_path_non_affine(path)
  File "/opt/miniconda2/lib/python2.7/site-packages/cartopy/mpl/geoaxes.py",
line 183, in transform_path_non_affine
    getattr(self, 'force_path_ccw', False))
  File "/opt/miniconda2/lib/python2.7/site-packages/cartopy/mpl/patch.py",
line 192, in path_to_geos
    geom = sgeom.Polygon(path_verts[:-1, :])
  File
"/opt/miniconda2/lib/python2.7/site-packages/shapely/geometry/polygon.py",
line 229, in __init__
    self._geom, self._ndim = geos_polygon_from_py(shell, holes)
  File
"/opt/miniconda2/lib/python2.7/site-packages/shapely/geometry/polygon.py",
line 508, in geos_polygon_from_py
    geos_shell, ndim = geos_linearring_from_py(shell)
  File "shapely/speedups/_speedups.pyx", line 214, in
shapely.speedups._speedups.geos_linearring_from_py
(shapely/speedups/_speedups.c:4017)
ValueError: A LinearRing must have at least 3 coordinate tuples
```

I am using current conda-forge versions of hopefully everything on python2.7.  I'm digging more at this and will generate a minimal reproducer as I can isolate what is happening.  I believe it is some boundary gradient issue as extending the plot bounds will typically stop this problem.

**Comment by akrherz** [comment]
The "bad" path segment has the following characteristics
```
path_verts = [[-90.7         43.11909091]
 [-90.7         43.11909091]
 [-90.7         43.11909091]]
path_codes = [ 1  2 79]
```
So `sgeom.Polygon(path_verts[:-1, :])` ends up creating a two point polygon

**Comment by akrherz** [comment]
Applying this patch
```
diff --git a/lib/cartopy/mpl/patch.py b/lib/cartopy/mpl/patch.py
index d8876ec..f64aab4 100644
--- a/lib/cartopy/mpl/patch.py
+++ b/lib/cartopy/mpl/patch.py
@@ -188,10 +188,7 @@ def path_to_geos(path, force_ccw=False):
         elif (path_verts.shape[0] > 2 and
                 (path_codes[-1] == Path.CLOSEPOLY or
                  verts_same_as_first[-1])):
-            if path_codes[-1] == Path.CLOSEPOLY:
-                geom = sgeom.Polygon(path_verts[:-1, :])
-            else:
-                geom = sgeom.Polygon(path_verts)
+            geom = sgeom.Polygon(path_verts)
         else:
             geom = sgeom.LineString(path_verts)
```
makes logical sense to me, but then still emits warnings
```
IllegalArgumentException: Invalid number of points in LinearRing found 3 - must be 0 or >= 4
Shell is not a LinearRing
```
I wonder if the bug is with the clipping done by matplotlib (v2.0.2).  Will punt for the moment

**Comment by akrherz** [comment]
Another similar error that I often see from my cartopy scripts
```text
Too few points in geometry component at or near point -10230261.203901839
5099199.1645516697
```
Boggling this one as well here as it is still emitted even with the patch I proposed .

**Comment by QuLogic** [comment]
@sgillies I think this requirement for the length of a `LinearRing` is not really documented, is it? I guess it makes sense, though.

**Comment by sgillies** [comment]
@QuLogic it's not adequately documented, I agree: http://toblerity.org/shapely/shapely.geometry.html#shapely.geometry.polygon.LinearRing.

**Comment by QuLogic** [comment]
Fixed by #885.

In the future, you can use "Fixes #..." in the PR description to auto-close relevant issues.
